packages/core/src/interpreter.ts

```diff
@@ -809,73 +802,68 @@ export class Interpreter<
 
         break;
       case actionTypes.start: {
-        const activity = (action as ActivityActionObject<TContext, TEvent>)
-          .actor as InvokeDefinition<TContext, TEvent>;
-
-        // If the activity will be stopped right after it's started
-        // (such as in transient states)
-        // don't bother starting the activity.
-        // if (!this.state.activities[activity.type]) {
-        //   break;
-        // }
+        const actorDef = (action as InvokeActionObject<TContext, TEvent>).def;
 
         // Invoked services
-        if (activity.type === ActionTypes.Invoke) {
-          const serviceCreator:
-            | ServiceConfig<TContext, TEvent>
-            | undefined = this.machine.options.services
-            ? this.machine.options.services[activity.src]
-            : undefined;
 
-          const { id, data } = activity;
+        let spawnableCreator = isString(actorDef.src)
+          ? this.machine.options.services
+            ? this.machine.options.services[actorDef.src]
+            : undefined
+          : actorDef.src;
 
-          const autoForward =
-            'autoForward' in activity
-              ? activity.autoForward
-              : !!activity.forward;
+        if (isMachineNode(spawnableCreator)) {
+          spawnableCreator = spawnMachine(spawnableCreator);
+        }
 
-          if (!serviceCreator) {
-            // tslint:disable-next-line:no-console
-            if (!IS_PRODUCTION) {
-              warn(
-                false,
-                `No service found for invocation '${activity.src}' in machine '${this.machine.id}'.`
-              );
-            }
-            return;
+        const { id, data } = actorDef;
+
+        if (!spawnableCreator) {
+          // tslint:disable-next-line:no-console
+          if (!IS_PRODUCTION) {
+            warn(
+              false,
+              `No service found for invocation '${actorDef.src}' in machine '${this.machine.id}'.`
+            );
           }
+          return;
+        }
 
-          const actor = serviceCreator(context, _event.data, {
+        const spawnable = (spawnableCreator as ActorCreator<TContext, TEvent>)(
+          context,
+          _event.data,
+          {
             parent: this as any,
             id,
             data,
             _event
-          });
-
-          if (autoForward) {
-            this.forwardTo.add(id);
           }
+        );
 
-          this.children.set(id, actor);
-
-          const childIndex = this.state.children.findIndex(
-            child => child.id === id
-          );
+        const actor = actorFrom(spawnable, id, this as Actor);
 
-          this.state.children[childIndex] = actor;
+        // Mutate the deferred actor reference
+        if (action.actor) {
+          Object.assign(action.actor, actor);
```

Comment:
for what part of the codebase or the consuming code this is relevant? mutating the action seems like a strong anti-pattern, so I'm wondering if we could think of a different design to avoid this. I don't understand yet the reasons why this is done though

Reply:
The reason for this is that at first, the actor is "deferred" - it's just a null actor. You can think of this as the result of:

```js
machine.transition('STATE', 'EVENT');
// => resulting state spawned an actor!
```

Obviously, we don't want an actor to actually be spawned (yet), so it's "deferred" until the interpreter decides to make it alive.

This takes inspiration from Akka:

> This particular main Actor is created using Actor.deferred, which is like a factory for a behavior. Creation of the behavior instance is deferred until the actor is started, as opposed to Actor.immutable that creates the behavior instance immediately before the actor is running.

https://doc.akka.io/docs/akka/2.5.5/scala/typed.html

And also it solves the problem of spawning actors in an initial state, since you can now do that without fear of the actor unintentionally being started before the service is started.

Reply:
Right - I understand that we want `machine.transition` to be pure and thus we cannot actually spawn an actor within it. I'm more interested in why exactly we want to represent them immediately in the returned State value over just returning "pure" actions and having Interpreter to actually execute them (and thus spawn actors).

Reply:
Because they need to be present in the `context`.

**State-less machine:**
```js
machine.transition(someState, someEvent);
// State {
//   ...
//   context: { actor: deferredActor }
// }
```

**Interpreted machine:**
```js
service.onTransition(/* ... */)
// State {
//   ...
//   context: { actor: liveActor }
// }
```

**Persisted state:**
```js
// The information is used to restore the actor
JSON.stringify(state);
// State {
//   ...
//   context: { actor: { id, meta } }
// }
```

We do not want `actor: undefined` or `actor: null` because:

1. it's not helpful for testing
2. it loses ability to be rehydrated
3. it can't be linked to in `state.children`

Actually, that point, linking an actor to `state.children` is the most important point - it's sub-optimal (if not very difficult/impossible) to do so if the `actor` is null first.

We want the state from `machine.transition(...)` to closely reflect `service.onTransition(...)` as much as possible.